File: 74. Search a 2D Matrix/74. Search a 2D Matrix.rb
# @param {Integer[][]} matrix
# @param {Integer} target
# @return {Boolean}
def search_matrix(matrix, target)
    return false if matrix.empty? || matrix[0].empty?
    
    m, n = matrix.length, matrix[0].length
    left, right = 0, m * n - 1
    
    while left <= right
        mid = left + (right - left) / 2
        mid_value = matrix[mid / n][mid % n]
        
        if mid_value == target
            return true
        elsif mid_value < target
            left = mid + 1
        else
            right = mid - 1
        end
    end
    
    false
end

File: 74. Search a 2D Matrix/74. Search a 2D Matrix.js
const searchMatrix = (matrix, target) => {
    if (matrix.length === 0 || matrix[0].length === 0) return false;

    const m = matrix.length;
    const n = matrix[0].length;

    let left = 0;
    let right = m * n - 1;

    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        const midValue = matrix[Math.floor(mid / n)][mid % n];

        if (midValue === target) {
            return true;
        } else if (midValue < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return false;
};

// Example usage
console.log(searchMatrix([[1,3,5,7],[10,11,16,20],[23,30,34,60]], 3)); // Output: true
console.log(searchMatrix([[1,3,5,7],[10,11,16,20],[23,30,34,60]], 13)); // Output: false